### PR TITLE
fix: esc() for 'raw' context

### LINKED ISSUE #8624: Bug: esc(..., 'raw')  fails on recursive data structures
### PHP Version

8.2

### CodeIgniter4 Version

4.4.6

### CodeIgniter4 Installation Method

Composer (as dependency to an existing project)

### Which operating systems have you tested for this bug?

Linux

### Which server did you use?

apache

### Database

_No response_

### What happened?

In the process of upgrading from CI3 -> 4, I encountered a problem with a `view()` that uses a fairly complex `$data` structure including &references between elements. Even though the `view()` helper calls `setData()` with `$context = 'raw'`, this still ends up in infinite recursion because `esc()` iterates through the fields of the array and calls `esc(..., 'raw')` on them. This whole operation is completely unnecessary because in the end it does not escape anything. However, because of the self-references, this walking of the array ends up in infinite recursion.

The `esc()` function should return `$data` immediately in the case of 'raw'. There's no need to walk the whole array tree just to build it anew.

### Steps to Reproduce

A non-realistic example, just to illustrate the point:
```php
$data = ['key' => []];
$data['key']['recur'] = &$data;
esc($data, 'raw');
```

### Expected Output

The exact same $data array, without iterating its values.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by Cleric-K**
**Description**
Fixes #8624

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by Cleric-K** [comment]
> Please sign all commits. See https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/signing.md

Sorry, I saw the requirement after PRing. It should be OK now.

File: system/Common.php [hunk 1]
```diff
@@ -435,13 +442,6 @@ function esc($data, string $context = 'html', ?string $encoding = null)
         if (is_string($data)) {
             $context = strtolower($context);
```
Comment (kenjis) [inline system/Common.php:443]: It seems this line also should be moved up.

Comment (Cleric-K) [reply system/Common.php]: Oh, right. Sorry. It should be OK now.

**Review by ddevsr — COMMENTED** [review]
Can you add tests for data array? recursive and non-recursive

**Comment by kenjis** [comment]
The following php-cs-fixer error is not related to this PR.
```
Files that were not fixed due to errors reported during fixing:
   1) /home/runner/work/CodeIgniter4/CodeIgniter4/app/Config/Logger.php
```
https://github.com/codeigniter4/CodeIgniter4/actions/runs/8327332402/job/22819194579?pr=8633

I sent a bug report: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/7896

**Comment by Cleric-K** [comment]
> Can you add tests for data array? recursive and non-recursive

Can you point me to an existing test for a function that I can use as an example? I'm not familiar with the CI project structure.

**Comment by kenjis** [comment]
See
https://github.com/codeigniter4/CodeIgniter4/blob/fd6d674274a332fe2d1e8ce2b1cad814cec934c5/tests/system/CommonFunctionsTest.php#L231-L248

**Comment by Cleric-K** [comment]
I added two tests. I don't have much experience with unit tests so please tell me if it seems right.

**Comment by kenjis** [comment]
@Cleric-K Please fix the coding style.
`composer cs-fix` will fix the errors.
See https://github.com/codeigniter4/CodeIgniter4/actions/runs/8343103783/job/22862006485?pr=8633

File: system/ThirdParty/Escaper/Escaper.php [hunk 2]
```diff
@@ -157,9 +157,21 @@ public function __construct(?string $encoding = null)
...
+        $this->jsMatcher       =
+            /** @param array<array-key, string> $matches */
+            function (array $matches): string {
+                return $this->jsMatcher($matches);
+            };
+        $this->cssMatcher      =
+            /** @param array<array-key, string> $matches */
+            function (array $matches): string {
+                return $this->cssMatcher($matches);
+            };
```
Comment (kenjis) [inline system/ThirdParty/Escaper/Escaper.php:174]: Sorry. please remove the change.
We don't upgrade Escaper in `develop` branch.

Comment (Cleric-K) [reply system/ThirdParty/Escaper/Escaper.php]: Oops, I haven't noticed how that got committed.

**Review by kenjis — APPROVED** [review]
LGTM!

**Comment by kenjis** [comment]
@Cleric-K Thank you!

**Pull request merged**